src/ast/nodes/shared/Expression.ts

```diff
@@ -0,0 +1,74 @@
+import { WritableEntity } from '../../Entity';
+import { GenericNode, Node } from './Node';
+import { ObjectPath } from '../../variables/VariableReassignmentTracker';
+import CallOptions from '../../CallOptions';
+import ExecutionPathOptions from '../../ExecutionPathOptions';
+import { UNKNOWN_EXPRESSION, UNKNOWN_VALUE } from '../../values';
+
+export type PredicateFunction = (node: Expression) => boolean;
+export type SomeReturnExpressionCallback = (options: ExecutionPathOptions) => PredicateFunction;
+export type ForEachReturnExpressionCallback = (options: ExecutionPathOptions) => (node: Expression) => void
+
+export interface Expression extends WritableEntity {
+	/**
+	 * Executes the callback on each possible return expression when calling this node.
+	 */
+	forEachReturnExpressionWhenCalledAtPath (
+		path: ObjectPath,
+		callOptions: CallOptions,
+		callback: ForEachReturnExpressionCallback,
+		options: ExecutionPathOptions
+	): void;
+	getValue (): any;
+	hasEffectsWhenAccessedAtPath (path: ObjectPath, options: ExecutionPathOptions): boolean;
+	hasEffectsWhenCalledAtPath (path: ObjectPath, callOptions: CallOptions, options: ExecutionPathOptions): boolean;
+
+	/**
+	 * Should return true if some possible return expression when called at the given
+	 * path returns true.
+	 */
+	someReturnExpressionWhenCalledAtPath (
+		path: ObjectPath,
+		callOptions: CallOptions,
+		predicateFunction: SomeReturnExpressionCallback,
+		options: ExecutionPathOptions
+	): boolean
+}
+
+export interface ExpressionNode extends Expression, Node {}
```

Comment:
Hmm I'm not sure what there is both Expression and ExpressionNode. Is this because variables can be treated like Expression?